rfcs/20200817-jit-code-generation-for-aarch64/README.md

```diff
@@ -0,0 +1,256 @@
+for Aarch64 as Xbyak provides for x64. Therefore, the same functionality that 
+oneDNN achieves for x64 using Xbyak can be achieved for AArch64 from scratch 
+using Xbyak_aarch64.
+**This RFC introduces Xbyak_aarch64 into oneDNN and enables JIT code generation 
+for AArch64.**
+
+Ideally, all implementations written for x64 should be rewritten for AArch64 
+using Xbyak_aarch64 (Fig. Direct method) to obtain maximum performance of 
+AArch64.
+
+![Direct method](direct_method.png)
+
+However, this process takes a lot of man-hours. To solve this problem, 
+we have developed 
+[Xbyak_translator_aarch64](https://github.com/fujitsu/xbyak_translator_aarch64). 
+By using Xbyak_translator_aarch64 instead of Xbyak in JIT code generation 
+for x64, we can generate JIT code for AArch64 instead of JIT code for x64 
+(Fig. Indirect method). Since we can reuse jit_\*.(cpp|hpp) for x64 with just 
+a few changes for AArch64, and we can reduce our development effort for 
+AArch64 to a great extent. We call this as "Indirect method". So **this RFC 
+also introduces Xbyak_translator_aarch64 and 
+[xed](https://github.com/fujitsu/xed "xed") for AArch64.**
+
+![Indirect method](indirect_method.png)
+
+
+As a third method, it is also possible that we use x64's jit_\*.(cpp|hpp) 
+files and their translation to AArch64 as a baseline, and partially rewrite 
+them from Xbyak functions to Xbyak_aarch64 ones to achieve better performance 
+for AArch64. We call this method as "Mixed-method".
+
+In summary, this RFC allows us to generate JIT code for AArch64 in the 
+following three ways.
+- Direct method:Implement JIT code generation from scratch using Xbyak_aarch64
+- Indirect method:Generate JIT code for AArch64 from jit_\*.(cpp|hpp) 
+for x64 by Xbyak_translator_aarch64
+- Mixed method:Use jit_\*.(cpp|hpp) files for x64 as a baseline, tune them 
+partially with Xbyak_aarch64
+
+We applied these three methods to oneDNN code in v0.21 and confirms that 
+the JIT code is generated on AArch64, it passes tests/gtests/test_\*, 
+and it accelerates oneDNN processing as shows in the slide below.
+The source code is disclosed as 
+[https://github.com/fujitsu/dnnl_aarch64](https://github.com/fujitsu/dnnl_aarch64).
+
+
+![Performance evaluation of JIT](performance_evaluation.png)
+
+
+## Related RFCs and Issues
+
+- [https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20200401-cpu-dir-structure](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20200401-cpu-dir-structure)
+- [https://github.com/oneapi-src/oneDNN/pull/795](https://github.com/oneapi-src/oneDNN/pull/795)
+- [https://github.com/oneapi-src/oneDNN/issues/804](https://github.com/oneapi-src/oneDNN/issues/804)
+
+
+## Proposal
+
+- Adding xbyak_translator_aarch64 directories to src/cpu/aarch64
+- Adding a dependency on xed at the build stage
+- Adding primitive entry to src/cpu/cpu_\*_list.cpp
+
+### Adding xbyak_translator_aarch64 directories to src/cpu/aarch64
```

Comment:
So, as far as I understand, the _translator_ is something you use offline, to convert x64 kernels to sve / simd&fp ones. Moreover, some of the kernels were then adjusted or further tuned.
If so, what is the reason to put it inside oneDNN?

To me from the library perspective it makes sense to:
- Put `xbyak_aarch64` to `src/cpu/aarch64/xbyak_aarch64` just like we include the original `xbyak` in the source tree
- Put sve and simd&fp implementations to `src/cpu/aarch64`
- Mention (maybe put a `README.md` in `src/cpu/aarch64`) that sve and simd&fp implementations are obtained by translating the code using the translator, and give a link to it.

Outside of the library:
- Keep maintaining `xbyak_aarch64`. It seems like a very useful standalone project. From time to time pull the changes from the upstream to oneDNN copy of it. We could also consider using git submodules, if you want.
- Keep maintaining the translator separately form the library and `xbyak_aarch64`. It is a standalone tool that could be used to add new kernels to the library, but is actually **not required** for building the library per se.

Reply:
- Xbyak_translator_aarch64 is not a tool that rewrites source code for x64 written by Xbyak, but a tool that rewrites x64 binaries made by oneDNN at runtime into AArch64 binaries, which is similar to QEMU, so we need to use the translator in combination with oneDNN at runtime.
- Thank you for your suggestion of submodule. There are both advantages and disadvantages to making them sub-modules, but we think it's easier for code management to put the source code directly within oneDNN, just like xbyak.

Reply:
![code_generation_and_usage](https://user-images.githubusercontent.com/27798870/90778981-44ca3e00-e338-11ea-873c-c7ab81205c02.png)
The word QEMU can be confusing.
Because oneDNN generate a variety of code such as layer fusion, post operation, optimization of loop body length and etc at runtime,  translation must also be done at runtime.

Reply:
![code_generation_and_usage](https://user-images.githubusercontent.com/27798870/90779583-f4071500-e338-11ea-874f-ff9060cb2d04.png)
The word QEMU can be confusing.
Because oneDNN generate a variety of code such as layer fusion, post operation, optimization of loop body length and etc at runtime, translation must also be done at runtime.

Reply:
Hi @kawakami-k,

First of all, I am sorry for the delay with response. I was out for some time.

Thank you for you explanation!

I still feel I might have missed something so please bear with me.

There are 3 essential components that I see in the proposal (just laid them in the more compact form, sorry for repeating the RFC in a sense):
1. `Xbyak_aarch64`. This one is clear to me -- an Aarch64 analogue Xbyak from Mitsunari-san;
2. `Xbyak_translator_aarch64`. The tool that allows translating the x64 Xbyak code to Aarch64.
    1. The translator exposes the modified `Xbyak`, so that the code (say `jit_avx512_conv_kernel`) continue using the x64 mnemonics, but in the end would actually generate Aarch64 instructions.
    2. The translator uses `Xbyak_aarch64` to generate Aarch64 instructions.
    3. The translator uses slightly modified `xed` to decode the instructions.
    4. I also noticed [`mbuild`](https://github.com/fujitsu/xbyak_translator_aarch64/tree/fjdev/translator/third_party) which seems to be using to build `xed`.
3. The oneDNN extensions:
    1. Former x64 code (from version 0.21) that is checked and maybe **slightly** adjusted to work with `Xbyak_translator_aarch64`. This code is mostly written using x64 mnemonics.
    2. Completely new SVE code (e.g. [this one](https://github.com/fujitsu/dnnl_aarch64/blob/release_base_0.21/src/cpu/jit_sve_conv_kernel.cpp)) that was based on the corresponding x64 code, but is now completely re-written using directly `Xbyak_aarch64`.

Now few extra questions I have:
1. Mostly to confirm: do I understand correctly that from `Xbyak_translator_aarch64` you plan to put only `translator/include/` directory in oneDNN; Xed will be an external dependency and tests/utils/examples won't go either? This will roughly lead to +6 MB of the source file size.
2. In the RFC you mentioned that only `*_aarch64_sve_*` implementations are added (see [1] below). I assume that these implementations not only correspond to 3.2 above, but also correspond to 3.1. In other words, in your fork you sometimes re-used x64 directly when possible (e.g. [here](https://github.com/fujitsu/dnnl_aarch64/blob/release_base_0.21/src/cpu/cpu_engine.cpp#L299)), in addition to adding SVE based implementations directly (e.g. [here](https://github.com/fujitsu/dnnl_aarch64/blob/release_base_0.21/src/cpu/cpu_engine.cpp#L169)). The question is: how would you integrate the code that was previously reused/shared with x64?
Let's take an example of the aforementioned `jit_uni_eltwise_fwd_t<avx512_common, f32>`. Do you plan to copy this implementation from the fork into `src/cpu/aarch64/` or do you plan to port the necessary changes to the current `src/cpu/x64/jit_uni_eltwise*.*` files and enable this _common_ implementation for both X64 and AARCH64? I assume the former, but wanted to double check.
    - I was just surprised that in the RFC there were only `_sve_512_` added primitives. So wanted to check how the others will look like (e.g. this uni eltiwse).

---

[1]

```
        CPU_INSTANCE_X64(jit_sse41_dw_convolution_fwd_t)
        CPU_INSTANCE_X64(jit_sse41_1x1_convolution_fwd_t)
        CPU_INSTANCE_X64(jit_avx2_convolution_fwd_t)
        CPU_INSTANCE_X64(jit_sse41_convolution_fwd_t)
        CPU_INSTANCE_AARCH64(jit_aarch64_sve_512_common_dw_convolution_fwd_t)         <- Added
        CPU_INSTANCE_AARCH64(jit_aarch64_sve_512_common_1x1_convolution_fwd_f32_t)    <- Added
        CPU_INSTANCE_AARCH64(jit_aarch64_sve_512_core_f32_wino_conv_2x3_fwd_t)        <- Added
        CPU_INSTANCE_AARCH64(jit_aarch64_sve_512_core_f32_wino_conv_4x3_fwd_t)        <- Added
        CPU_INSTANCE_AARCH64(jit_aarch64_sve_512_common_convolution_winograd_fwd_t)   <- Added
        CPU_INSTANCE_AARCH64(jit_aarch64_sve_512_common_convolution_fwd_t<f32>)       <- Added
```

Reply:
Thanks for summarizing our RFC's main points clearly.
I think your summary is an accurate and concise description of this RFC.

1. 
I'm sorry I didn't explicitly state what files are added to oneDNN by this RFC.
The following tree shows what we'd like to add to `src/cpu/aarch64/xbayk_translator_aarch64`.

```
src/cpu/arch64/xbyak_translator_aarch64
|-- translator
|   |-- include
|   |   `-- xbyak_translator_for_aarch64
|   |       |-- instructions
|   |       |   |-- adc.h
|   |       |   |-- add.h
|   |       |   |-- (other header files for various x64 instructions)
|   |       |-- xbyak_translator.h
|   |       |-- xbyak_translator_debug.h
|   |       |-- xbyak_translator_inc.h
|   |       |-- xbyak_translator_switch.h
|   |       |-- xbyak_translator_utils.h
|   |       `-- xed-examples-util_aarch64.h
|   |-- third_party
|   |   |-- xbyak_aarch64
|   |   |   `-- xbyak_aarch64
|   |   |       |-- xbyak_aarch64.h
|   |   |       |-- xbyak_aarch64_adr.h
|   |   |       |-- xbyak_aarch64_code_array.h
|   |   |       |-- xbyak_aarch64_err.h
|   |   |       |-- xbyak_aarch64_gen.h
|   |   |       |-- xbyak_aarch64_inner.h
|   |   |       |-- xbyak_aarch64_label.h
|   |   |       |-- xbyak_aarch64_meta_mnemonic.h
|   |   |       |-- xbyak_aarch64_mnemonic.h
|   |   |       `-- xbyak_aarch64_reg.h
|-- xbyak  (we modified Xbyak a bit to hook its process for translation.)
    |-- xbyak.h
    |-- xbyak_bin2hex.h
    |-- xbyak_mnemonic.h
    `-- xbyak_util.h
```

Since XED is an independent external library, it doesn't have to be inside oneDNN.
As well as https://github.com/diaena/oneDNN/tree/arm-rfc/rfcs/20200731-Using-Arm-Compute-Library-kernels-on-AArch64,
additional CMake modules will be needed (FindXED.cmake, xed.cmake) which will only impact AArch64 builds.
Build steps for AArch64 become as follows.
- Build xed library preliminary
- Set environment variable `export XED_ROOT_DIR=/path/to/XedLibrary;`
- Execute `cmake` for oneDNN
- Execute `make` for oneDNN

`Xbyak_translator_aarch64` and `Xbyak_aarch64` originally contain a few MB of test code,
but that will be removed as well.


2.
Our plan is the former. Initially, we will reuse the x64 implementation, but we expect to gradually increase the differences as we proceed optimization using Xbyak_aarch64 directly. 
In that case, I think it would be difficult to keep it to a single source code.

We are currently developing protoype code in
https://github.com/fujitsu/oneDNN, which was forked on July 25.
We copied `src/cpu/aarch64/jit_uni_*.(cpp|hpp)` from `src/cpu/x64/`.
We just replaced `x64` in the files to `aarch64` and commented out some lines.
Some native implementation with Xbyak_aarch64 is also included such as `src/cpu/aarch64/jit_aarch64_sve_512_*.(cpp|hpp).
(Note: This prototype does not reflect the discussion in this RFC.)

At this time, it can be build and some JIT code can be generated, but some tests fail and `Xbyak_translator_aarch64` can't translate some x64 instructions, which are additionally used by the current master branch compared to v0.21.2. We are developing the translator and the prototype to extend support instructions and primitives.

Reply:
I see, thank you very much for you patient explanation!